executor/executor.go

```diff
@@ -32,7 +35,21 @@ type Executor struct {
 	done      bool
 	completed int
 	tasks     map[int]*task
-	edges     map[string]int
+	edges     map[string]*data
+}
+
+// invariant: [data] holds either 1 Allocate/Write or
+// a set of Reads, but never both
+//
+// invariant: [waiter] is made every time when setting
+// Allocate/Write or when adding the first Read. [waiter]
+// is closed only when [blockers] == 0
+type data struct {
+	allocateWrite int
+	reads         set.Set[int]
```

Comment:
Something isn't adding up to me. It doesn't look like you're ever reading `allocateWrite` or `reads`, just adding to the `reads` set, or setting `allocateWrite` to an ID. Could you not accomplish the same thing with a bool? It's either blocking on reads (the number of which corresponds to `blockers`) or blocking on a single write. 

I'm not sure if you have plans to extend this, but you could definitely enforce your invariants by code via 
```suggestion
	isWrite bool
```
like I said, I could be missing something though